### ISSUE TITLE
Questions about training details for BB and surface code experiments

### ISSUE THREAD

**Issue by Patrickz1201** [arshpreetmaan/astra#1]
Hello,

Thank you for releasing the code for the paper "Machine learning message-passing for the scalable decoding of QLDPC codes". The implementation has been very helpful for understanding the method.

I am currently trying to reproduce the results reported in the paper using the public repository, and I encountered several issues where I would appreciate some clarification regarding the training setup.

### 1. BB codes (distance 12 and 18)

Using the parameters described in the paper, I was able to successfully reproduce the BB code results for **distance 6**.However, I was not able to reproduce the reported performance for **distance 12 and distance 18**.I was wondering if you could clarify:

- From which **training epoch** the training for d12 and d18 were started?
- Were there any **additional training hyperparameters** used that are not explicitly listed in the paper (e.g., batch size, learning rate, weight decay, scheduler, etc.)?
- Were the models trained using **the same training procedure as d6**, or were there any modifications for the larger distances?

---

### 2. Surface codes

When trying to reproduce the **surface code experiments**, I encountered two issues:

**(1) Memory issue**

When training the model for **distance 6**, the training process runs out of GPU memory.

My current setup:
- GPU: NVIDIA A800
- number of iterations: 30
- batch size: 16
- node features: 500
- edge features: 500
- msg_net_size: 512

I noticed in the paper that some experiments were trained on an **RTX 3080**, so I was wondering whether:

- mixed precision (FP16) was used during training,
- the batch size was smaller than 16,
- or other techniques (e.g., gradient accumulation or checkpointing) were used.

**(2) Performance gap for distance 3**

For **distance 3**, the training runs successfully, but the final decoding performance is noticeably worse than the results reported in the paper.

I was wondering if there are any important training details (e.g., number of iterations during training vs. inference, batch size, lr) that might affect the final performance.

---

Any clarification would be greatly appreciated. Thank you again for making the code and results available.

Best regards

**Comment by arshpreetmaan** [comment]
Hello,

Thank you for your feedback. I'm updating the instructions in the codebase for easier reproducibility.

Point-by-point clarification to the comments:

**1. BB codes:**

- The maximum training epoch is listed in the codebase. We train until the model stops improving or archives LER sufficiently below the baseline decoder (bposd in this case)
- Most of them are listed in the Table 1 in the paper or in the training procedure paragraph. Rest of them (weight decay etc) are in the codebase. Batch size depends on the memory of the GPU used. User needs to adjust is based on that. 
- In general, we sampled new training data for each epoch. This ensures that model get to see a lot of samples and improves the learning quality. This option is available in the codebase and mentioned in the paper (but it was not being used by default -- user need to uncomment those lines of code)  

 **2. Surface codes:**

1. Memory issue: Typically i was using maximum which fills the memory for a given distance. For d3, batch size of 256 fits will within the memory of 3080 for the mentioned hyperparameters. Yes, mixed precision is being used as default in the codebase. 
2. I was quickly able to reproduce the results (in few runs) after cloning the repo (see plot below). Usually increasing the number of iteration during inference improves the convergence, thus improving the LERs. Typically, we used 100 iterations for surface code and 200 for BB code. Moreover, as mentioned in the paper, all the curves in the plots were generated by a model trained on d11 for surface code and d18 for BB codes respectively.

<img width="848" height="487" alt="Image" src="https://github.com/user-attachments/assets/7d9a9b66-9211-4318-96f3-6f0a49f24f1e" />
Here the plot shows, the new astra d3,d5,d7 models which were trained in past days after cloning this repo. They are being compared with the results in the paper. The original d3,d5,d7 are actually d11-d3,d11-d5 and d11-d7 respectively i.e. generated using model trained on d11. 

\
I hope this clarifies the issues. Please feel free to reach out if you have more questions.
